### PR TITLE
..../---/.-- -.-./---/..-/.-../-.. .--/. ..-./---/.-./--././- -/..../…

### PR DESCRIPTION
**Pull request by minad**
…./.../. ...-/./.-./-.-- ../--/.--./---/.-./-/.-/-./- .-/-.-./-/../---/-./.../..--..

### CODE REVIEW

**Comment by oantolin** [comment]
I mean, I did add a binding for `rot13-region` (which I do actually use sometimes).

**Comment by minad** [comment]
More seriously - we could introduce an embark-encode-map (similar to the sort map) which could have such encodings and maybe even some practical ones (url encode for example).

**Comment by oantolin** [comment]
That's a good idea. We have no reason to avoid prefixes anymore, now that the indicators can handle them correctly.

**Comment by oantolin** [comment]
Are there some other encoding commands it would be useful to have here? I guess at least base64 and url-encode-url (which requires a small wrapper, I guess). There is something called base64url as well.

**Comment by minad** [comment]
There are also `format-encode-region` and `format-decode-region` which ask the user for the desired format. See also `format-alist`.

**Comment by oantolin** [comment]
If we really wanted to avoid shadowing `embark-collect-snapshot` and `embark-export` we could bind `whitespace-cleanup-region` to `SPC` (shadowing `mark` which is useless for region targets!), and that leaves `s` open for the sort map. We could bind the encode map to `>`. What do you think?

**Comment by minad** [comment]
> What do you think?

Sounds good! The base64 actions are a bit questionable imo since they are more useful when operating on binary.

**Comment by oantolin** [comment]
Of course! Let's not add base64, then.

**Comment by minad** [comment]
Ah, I forgot. "e" is eval, which is important!

**Comment by minad** [comment]
Okay, from my side this is ready.

**Comment by oantolin** [comment]
I think I might add a couple of more already before merging. I also want to change the binding of `reverse-region` to lowercase `r` which is now free since we moved `rot13-region` under the prefix.

**Comment by minad** [comment]
Sure, please do! Also the `ispell-word` binding is questionable since there is `M-$`.

**Comment by oantolin** [comment]
> Sure, please do! Also the `ispell-word` binding is questionable since there is `M-$`.

Let the which-key users complain about clutter. 🙃

**Comment by oantolin** [comment]
Should the url-encode action replace the buffer text or save the encoded version on the kill-ring?

**Comment by minad** [comment]
> Should the url-encode action replace the buffer text or save the encoded version on the kill-ring?

I think the commands which encode/decode should directly manipulate the buffer. For non invertible commands it makes more sense to save to the kill ring, like the hash commands do.

**Comment by oantolin** [comment]
I agree, I guess the use case is simply writing a URL: you write it with spaces and whatnot, mark it, encode it.

File: embark.el [hunk 1]
```diff
@@ -3598,7 +3598,7 @@ and leaves the point to the left of it."
   ("m" embark-hash-md5)
   ("1" embark-hash-sha1)
   ("2" embark-hash-sha256)
-  ("3" embark-hash-sha356)
+  ("3" embark-hash-sha384)
```
Comment (minad) [inline embark.el:3601]: Oops :laughing:

**Comment by minad** [comment]
Yay, thanks for merging. Is there also a decode-url command? I haven't found such a function, but I didn't dig deeply. Eww must have something like this?

**Comment by oantolin** [comment]
There is a function but it's called url-unhexify-string. I added the corresponding action. bc460d9654c1728b7f8ffd5df8939c1f0054c32f

**Comment by oantolin** [comment]
Thanks for doing this. I think it's worth it even if just for publicizing that Emacs has all those hash functions. 😃

**Comment by minad** [comment]
You are welcome! Yes, I like Embark also for its exploratory aspect.

**Comment by oantolin** [comment]
I USED BOTH SHA256 AND Base64 TODAY!

**Comment by minad** [comment]
Unbelievable! I hope you could use it multiple times repeatedly such that you *really* profited from the nice Embark integration. I didn't even remember the keybinding, but > is a nice one!

**Pull request merged**